File: 2024/day02/main.py
#!/usr/bin/env python
"""
Red-Nosed Reports
=================

# Example input:
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9

Part 1: 2 safe reports
Part 2: 4 tolerable reports
"""
# Standard library
from collections.abc import Sequence
import itertools
import logging
import time

# 1st party
import aoc_utils as aoc

log = logging.getLogger('AoC')

################################################################################
def main() -> None:
    # Boilerplate
    args = aoc.parse_args()
    aoc.configure_logging(args.log_level)
    text = aoc.read_input(args.input)

    # Solution
    reports = [[int(n) for n in l.split()] for l in text.splitlines()]

    # Part 1
    start   = time.perf_counter()
    n       = sum(map(is_safe_report, reports))
    elapsed = time.perf_counter() - start
    print('Part 1:', n, f'[{elapsed*1000:.3f}ms]')

    # Part 2
    start   = time.perf_counter()
    n = sum(map(is_tolerable_report, reports))
    elapsed = time.perf_counter() - start
    print('Part 2:', n, f'[{elapsed*1000:.3f}ms]')

################################################################################
def is_safe_report(nums: Sequence[int]) -> bool:
    if len(nums) <= 1:
        return True
    is_increasing = nums[0] < nums[1]
    for x,y in itertools.pairwise(nums):
        if not (1 <= abs(y-x) <= 3):
            return False
        if (x<y) != is_increasing:
            return False
    return True

def is_tolerable_report(nums: Sequence[int]) -> bool:
    if len(nums) <= 2:
        return True
    return (
        _is_tolerable_report_increasing(nums)
        or
        _is_tolerable_report_increasing(nums[::-1])
    )

def _is_tolerable_report_increasing(nums: Sequence[int]) -> bool:
    '''
    Examples
    ========
    >>> _is_tolerable_report_increasing([1,2,4,7])
    True
    >>> _is_tolerable_report_increasing([1,2,3,9])
    True
    >>> _is_tolerable_report_increasing([9,1,2,3])
    True
    >>> _is_tolerable_report_increasing([9,1,2,9])
    False
    '''
    found_bad_level = False
    is_safe = lambda x, y: 1 <= y-x <= 3
    i = 0
    while i+1 < len(nums):
        if is_safe(nums[i], nums[i+1]):
            i += 1
            continue
        elif found_bad_level:
            return False
        elif i+2 == len(nums):
            return True
        found_bad_level = True

        # Determine if i or i+1 should be dropped
        if is_safe(nums[i], nums[i+2]): # drop i+1
            i += 2
            continue
        elif i == 0 and is_safe(nums[i+1], nums[i+2]): # drop i
            i += 2
            continue
        elif i != 0 and is_safe(nums[i-1], nums[i+1]): # drop i
            i += 1
            continue
        return False
    return True


if __name__ == "__main__":
    main()

File: 2024/day02/main.cpp

// 1st party
#include "run.hpp"
#include "aoc_utils.hpp"

// Standard library
#include <algorithm>
#include <concepts>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <ranges>
#include <sstream>
#include <vector>

// Aliases
namespace fs     = std::filesystem;
namespace aoc    = aoc_utils;
namespace ranges = std::ranges;
namespace views  = std::views;

////////////////////////////////////////////////////////////////////////////////
// Part 1
constexpr bool is_safe_level(int x, int y, bool is_increasing) {
    const int diff {std::abs(y-x)};
    return 1 <= diff && diff <= 3 && (x<y) == is_increasing;
}

template<ranges::random_access_range R>
bool is_safe_report(const R& nums) {
    return ranges::all_of(
        nums | views::pairwise,
        [is_increasing = nums[0] < nums[1]]
        (const auto pair){
            const auto [x, y] = pair;
            return is_safe_level(x, y, is_increasing);
        }
    );

    // Option 1: Procedural
    // if (nums.size() <= 1) {
    //     return true;
    // }
    // const bool is_increasing = nums[0] < nums[1];
    // for (size_t i=0; i < nums.size()-1; ++i) {
    //     if (!is_safe_level(nums[i], nums[i+1], is_increasing)) {
    //         return false;
    //     }
    // }

    return true;
}

////////////////////////////////////////////////////////////////////////////////
// Part 2
template<ranges::random_access_range R>
requires std::integral<std::ranges::range_value_t<R>>
bool _is_tolerable_report_increasing(const R& nums) {
    using T       = std::ranges::range_value_t<R>;
    using diff_t  = std::ranges::range_difference_t<R>;
    const auto sz = [](const diff_t i){
        return static_cast<std::ranges::range_size_t<R>>(i);
    };

    auto found_bad_level = false;
    diff_t i = 0;
    const auto begin = std::ranges::begin(nums);
    const auto is_safe = [](const T x, const T y) {return x+1 <= y && y <= x+3;};

    while (sz(i+1) < nums.size()) {
        if (is_safe(begin[i], begin[i+1])) {
            i += 1;
            continue;
        } else if (found_bad_level) {
            return false;
        } else if (sz(i+2) == nums.size()) {
            return true;
        }
        found_bad_level = true;

        // Determine if i or i+1 should be dropped
        if (is_safe(begin[i], begin[i+2])) { // drop i+1
            i += 2;
            continue;
        } else if (i == 0 && is_safe(begin[i+1], begin[i+2])) { // drop i
            i += 2;
            continue;
        } else if ( i != 0 && is_safe(begin[i-1], begin[i+1])) { // drop i
            i += 1;
            continue;
        }
        return false;
    }
    return true;
}

template<ranges::random_access_range R>
bool is_tolerable_report(const R& nums) {
    if (nums.size() <= 1) {
        return true;
    }
    return (
        _is_tolerable_report_increasing(nums)
        ||
        _is_tolerable_report_increasing(nums | views::reverse)
    );
}

////////////////////////////////////////////////////////////////////////////////
using Report = std::vector<int>;
using InputType = std::vector<Report>;
auto read_data(const fs::path &filepath) -> InputType {
    std::ifstream file(filepath);
    if (!file) {
        throw aoc::FileReadException(filepath);
    };

    std::string line;
    int val{0};
    std::vector<std::vector<int>> rows{};

    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::vector<int> row{};
        while (iss >> val) {
            row.push_back(val);
        }
        rows.push_back(row);
    }

    return rows;
}

////////////////////////////////////////////////////////////////////////////////
int main(int argc, char *argv[]) {
    const aoc::InputPathArgs args = aoc::parse_args(argc, argv);
    aoc::run<InputType>(
        args.input_path,
        read_data,
        /* part1 = */ [](const InputType& reports){ 
            return ranges::count_if(reports, is_safe_report<Report>);
        },
        /* part2 = */ [](const InputType& reports){ 
            return ranges::count_if(reports, is_tolerable_report<Report>);
        }
    );
}

